dart/common/Addon.h

```diff
@@ -0,0 +1,305 @@
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_COMMON_ADDON_H_
+#define DART_COMMON_ADDON_H_
+
+#include <string>
+
+#include "dart/common/Extensible.h"
+#include "dart/common/detail/NoOp.h"
+
+namespace dart {
+namespace common {
+
+class AddonManager;
+
+class Addon
+{
+public:
+
+  friend class AddonManager;
+
+  /// If your Addon has a State class, then that State class should inherit this
+  /// Addon::State class. This allows us to safely serialize, store, and clone
+  /// the states of arbitrary Addon extensions. If your Addon is stateless, then
+  /// you do not have to worry about extending this class, because
+  /// Addon::getState() will simply return a nullptr by default, which is taken
+  /// to indicate that it is stateless.
+  ///
+  /// The distinction between the State class and the Properties class is that
+  /// State will get stored in AddonManager::State whereas Properties will get
+  /// stored in AddonManager::Properties. Typically Properties are values that
+  /// only change rarely if ever, whereas State contains values that might
+  /// change as often as every time step.
+  class State : public Extensible<State> { };
+
+  /// Use the StateMixer class to easily create a State extension from an
+  /// existing class or struct.
+  template <class Mixin>
+  using StateMixer = ExtensibleMixer<State, Mixin>;
+
+  /// If your Addon has a Properties class, then it should inherit this
+  /// Addon::Properties class. This allows us to safely serialize, store, and
+  /// clone the properties of arbitrary Addon extensions. If your Addon has no
+  /// properties, then you do not have to worry about extending this class,
+  /// because Addon::getProperties() will simply return a nullptr by default,
+  /// which is taken to indicate that it has no properties.
+  ///
+  /// The distinction between the State class and the Properties class is that
+  /// State will get stored in AddonManager::State whereas Properties will get
+  /// stored in AddonManager::Properties. Typically Properties are values that
+  /// only change rarely if ever, whereas State contains values that might
+  /// change as often as every time step.
+  class Properties : public Extensible<Properties> { };
+
+  /// Use the PropertiesMixer class to easily create a Properties extension
+  /// from an existing class or struct.
+  template <class Mixin>
+  using PropertiesMixer = ExtensibleMixer<Properties, Mixin>;
+
+  /// Virtual destructor
+  virtual ~Addon() = default;
+
+  /// Clone this Addon into a new manager
+  virtual std::unique_ptr<Addon> cloneAddon(AddonManager* newManager) const = 0;
+
+  /// Set the State of this Addon. By default, this does nothing.
+  virtual void setAddonState(const std::unique_ptr<State>& otherState);
```

Comment:
This is a good point. Also, now I realize that the incorrect comment that @jslee02 pointed out was meant to apply to this. The current setup does not prevent the Addon from altering the contents of ``otherState``, which is a major flaw.

My motivation for passing in a ``unique_ptr`` here was to ensure that ``addon1->setAddonState(addon2->getAddonState()->clone());`` would always work. By default, the State of an Addon is a ``nullptr`` which is meant to indicate that it is stateless. This means that if pass in a dereferenced const reference, then ``addon1->setAddonState(*addon2->getAddonState()->clone())`` could segfault since we'd sometimes be dereferencing a ``nullptr``.

However, end users are **not** generally expected to be using this abstracted functionality. Instead, it should get handled automatically when they set/get the States or Properties of the class that has inherited ``AddonManager``. So it's not hard for us to put in a ``nullptr`` check before attempting to send in a ``State`` or ``Properties`` with the code which is actually meant to be using these functions.